core/quill.js

```diff
@@ -57,7 +57,12 @@ class Quill {
   }
 
   constructor(container, options = {}) {
-    this.options = expandConfig(container, options);
+    const document = options.document || window.document;
+    const selectionRoot = options.selectionRoot || document;
+
+    this.document = document;
+    this.selectionRoot = selectionRoot;
+    this.options = expandConfig(container, options, selectionRoot);
```

Comment:
With the new optional parameters, this is what the initialization in a webcomponent looks like (with simplification b/c of the HTMLElement lifecycle)


```
class QuillEditor extends HTMLElement {
  constructor() {
    super();
    const root = this.attachShadow({ mode: 'open' });
    root.innerHTML = `<div id="editor"></div>`;

    const editor=  root.getElementById('#editor');
    const quill = new Quill(editor, {
      modules: { .. },
      theme:  'snow',
      document: root.ownerDocument,
      selectionRoot: root
    });
  }
}
```